Firestore/src/support/secure_random.h

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright 2017 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_SRC_SUPPORT_SECURE_RANDOM_H_
+#define FIRESTORE_SRC_SUPPORT_SECURE_RANDOM_H_
+
+#include <limits>
+#include <stdint.h>
+
+namespace firestore {
+
+// A "secure" pseudorandom number generator, suitable for generating
+// unguessable identifiers. This exists because
+//
+//   * std::mt19937 is blazing fast but its outputs can be guessed once enough
+//     previous outputs have been observed.
+//   * std::random_device isn't guaranteed to come from a secure PRNG or be
+//     fast.
+//
+// The implementation satisfies the C++11 UniformRandomBitGenerator concept and
+// delegates to an implementation that generates high quality random values
+// quickly with periodic reseeding.
+class SecureRandom {
+ public:
+  // C++11 UniformRandomBitGenerator interface
+  using result_type = uint32_t;
+
+  static constexpr result_type min() {
+    return std::numeric_limits<result_type>::min();
+  }
+
+  static constexpr result_type max() {
+    return std::numeric_limits<result_type>::max();
+  }
+
+  result_type operator()();
```

Comment:
NB: the man page for arc4random recommends arc4random_uniform(max) over arc4random() for cases where the returned value is immediately mod'd by a number; i.e. "arc4random() % 10" (to avoid mod bias). Is that something we want to support here? (UniformRandomBitGenerator doesn't require it though, so if the point is to plug this into something that uses URBG interface, then adding this extra method won't help any.)